shapely/geometry/base.py

```diff
@@ -308,33 +321,33 @@ def geom_type(self):
     # ----------------------------------
 
     @property
-    def area(self):
-        """Unitless area of the geometry (float)"""
+    def area(self) -> float:
+        """Unitless area of the geometry"""
         return float(shapely.area(self))
 
-    def distance(self, other):
+    def distance(self, other: GeometryMaybeArrayLike) -> MaybeArray[float]:
```

Comment:
Would this be more precise?

```suggestion
    @typing.overload
    def distance(self, other: Geometry) -> float:
        ...
    @typing.overload
    def distance(self, other: ArrayLike[Geometry]) -> Array[float]:
        ...
    def distance(self, other: GeometryMaybeArrayLike) -> MaybeArray[float]:
```

Reply:
I'm not familiar with this syntax, is `def distance(self, other: GeometryMaybeArrayLike) -> MaybeArray[float]:` needed if you add the first two ?
But still, there are 27 functions like this in the unit alone, I'm afraid that this might add up to a lot of overhead. WDYT ?

Reply:
Yes, you need that last one which has the actual implementation. Personally, I think it's worth the overhead, but I can understand if the Shapely maintainers say it isn't. In that case, just typing it as `def distance(self, other: Geometry) -> float` would be my preferred option, leaving the array case as an advanced option, not supported by the typing. How often is the array case used? We don't use it in our project, but unsure about most of Shapely's users.

Reply:
> How often is the array case used? We don't use it in our project, but unsure about most of Shapely's users.

Probably the more common use of the array inputs/outputs is in the top level functions which accept array also as the first geometry (instead of `self`) so I guess it makes more sense there. 
In that case also I don't think that the [generics solution](https://github.com/shapely/shapely/issues/721#issuecomment-1409410818) would apply because there can be 2 `ScalarOrArrayLike` inputs (like `shapely.intersection`) and only if both are `scalar`s than the output is `scalar`, and otherwise it's an `array`.
So for the top level functions for sure we need to type the arrays, but for `base` I don't have a strong opinion.

Reply:
I think the object-oriented API of shapely is *mostly* used by people doing only scalar geometries. For these users, the typing doesn't really add value if you do (for instance) `distance(...) -> MaybeArray[float]`. A `cast` is then always necessary.

I am a bit torn on the solutions @parched suggested, but leaning towards the option of going all the way with the added `overload`. I'd like to hear the opinion of another maintainer on this @jorisvandenbossche  @sgillies . Would you accept added `@overload` decorated methods for a 100% correct typing?

Reply:
>  Would you accept added `@overload` decorated methods for a 100% correct typing?

Yes.